File: vi/codes/swift/chapter_stack_and_queue/stack.swift
/**
 * File: stack.swift
 * Created Time: 2023-01-09
 * Author: nuomi1 (nuomi1@qq.com)
 */

@main
enum Stack {
    /* Mã điều khiển */
    static func main() {
        /* Truy cập phần tử ở đỉnh ngăn xếp */
        // Swift không có lớp ngăn xếp tích hợp sẵn, có thể dùng Array làm ngăn xếp
        var stack: [Int] = []

        /* Đẩy các phần tử vào ngăn xếp */
        stack.append(1)
        stack.append(3)
        stack.append(2)
        stack.append(5)
        stack.append(4)
        print("Stack stack = \(stack)")

        /* Trả về danh sách để in ra */
        let peek = stack.last!
        print("Top element peek = \(peek)")

        /* Lấy phần tử ra khỏi ngăn xếp */
        let pop = stack.removeLast()
        print("Pop element pop = \(pop), after pop stack = \(stack)")

        /* Lấy độ dài của ngăn xếp */
        let size = stack.count
        print("Stack length size = \(size)")

        /* Kiểm tra rỗng */
        let isEmpty = stack.isEmpty
        print("Is stack empty = \(isEmpty)")
    }
}

File: vi/codes/cpp/chapter_stack_and_queue/stack.cpp
/**
 * File: stack.cpp
 * Created Time: 2022-11-28
 * Author: qualifier1024 (2539244001@qq.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    /* Khởi tạo ngăn xếp */
    stack<int> stack;

    /* Đẩy phần tử vào ngăn xếp */
    stack.push(1);
    stack.push(3);
    stack.push(2);
    stack.push(5);
    stack.push(4);
    cout << "Stack stack = ";
    printStack(stack);

    /* Truy cập phần tử ở đỉnh ngăn xếp */
    int top = stack.top();
    cout << "Stack top element top = " << top << endl;

    /* Lấy phần tử ra khỏi ngăn xếp */
    stack.pop(); // Không có giá trị trả về
    cout << "Pop element pop = " << top << ", after pop, stack = ";
    printStack(stack);

    /* Lấy độ dài của ngăn xếp */
    int size = stack.size();
    cout << "Stack length size = " << size << endl;

    /* Kiểm tra ngăn xếp có rỗng không */
    bool empty = stack.empty();
    cout << "Stack is empty = " << empty << endl;

    return 0;
}
